OpenGov-OpenData/ckanext-dcat
# [PERF-1596] Fix keep existing resources logic

## [PERF-1596](https://opengovinc.atlassian.net/browse/PERF-1596)

## Description

This PR does the following:
- For harvest jobs that have unmatched resources users can configure to keep them if `keep_existing_resources` is set to true. However, we should only keep resources if they have both a name and url.
- Changes the util function `parse_date_iso_format` to catch the generic ValueError exception instead of the specific ParserError exception
- Reorders the code of the configuration processors to be the order in which they are processed.
- Renames the config option `resources_order` to `resource_format_order` to be more descriptive
- Adds more tests for config processors

## Test Procedure
Requirements:
- The ckanext-harvest extension is installed
- The ckanext-dcat extension is installed
- Setup `/usr/lib/ckan/default/src/ckan/test-core.ini` with test DB config, [Testing CKAN](https://docs.ckan.org/en/2.7/contributing/test.html)

Install Instructions:
- Checkout PR branch in ckanext-dcat
```
cd /usr/lib/ckan/default/src/ckanext-dcat
git checkout jguo144/PERF-1596/2020-11-04/fix-keep-existing-resources-logic
```
- Ensure plugin list is configured correctly in CKAN ini config file
```
ckan.plugins = stats resource_proxy viewhelpers ags_fs_view ags_ms_view barchart basicgrid linechart piechart
  geo_view geojson_view c3charts datatables_view image_view officedocs_view pdf_view webpage_view
  odata datastore xloader extractor harvest datajson dcat dcat_rdf_harvester
  dcat_json_harvester dcat_json_interface structured_data opengov_harvester
  contact pages showcase theme_configurator og_theme security heroslideradmin
```
- Restart supervisor service for gather and fetch harvest queue
```
sudo service supervisor restart
```
- Restart CKAN in either terminal or apache
```
sudo service apache2 restart
```

Testing:
- Run nosetest
```
cd /usr/lib/ckan/default/src/ckanext-dcat
nosetests --ckan --with-pylons=test.ini ckanext/dcat/tests
```
- Create dcat json harvester job in CKAN UI for https://bostonopendata-boston.opendata.arcgis.com/data.json. Use the following harvest configuration.
```
{
  "resource_format_order": ["Esri REST", "HTML"],
  "keep_existing_resources": true
}
```
When the harvest job is complete check the dataset "Boston Street Segments Metadata". After the harvest job is ran again the "Esri Rest API" resource should not be duplicated.
